### DISCUSSION TITLE
API Code Examples?

### DISCUSSION THREAD

**Discussion by vne147** [lldpd/lldpd#487] [Q&A] [answered]
Hello!

I'm developing some C code that will use LLDP to query neighbors and specific information about them (interface, MAC, and IP).  I've used the documentation in lldpctl.h to write the code below, which seems to work well.  Right now it's just printing the IP and interfaces, but eventually I'll update it to probably push that info into a structure.

Next I'd like to register a callback to basically reset all the gathered information and re-query whenever any neighbor change is detected.  I've read the documentation on lldpctl_watch_callback2 and lldpctl_change_callback2, but I'm still having some trouble implementing it.  I'm probably not as proficient with C as I ought to be.

Is there any example code out there you can point me to that I can use as a template?  If it exists, I couldn't find it.

Thanks in advance for any help you can provide.
 
```
	lldpctl_conn_t *conn;
	lldpctl_atom_t *interfaces;
	lldpctl_atom_t *iface;
	lldpctl_atom_t *neighbors;
	lldpctl_atom_t *neigh;
	lldpctl_atom_t *port;
	lldpctl_atom_t *addresses;
	lldpctl_atom_t *adr;
	
	const char *ctlname;
	const char *ifname;
	const char *mgmt_ip;
	
	//Get default transport name
	ctlname = lldpctl_get_default_transport();
	
	//Make connection to LLDP
	conn = lldpctl_new_name(ctlname,NULL,NULL,NULL);
	
	//Get list of interfaces
	interfaces = lldpctl_get_interfaces(conn);	
	
	//Iterate over each interface
	lldpctl_atom_foreach(interfaces,iface){
		
		ifname = lldpctl_atom_get_str(iface,lldpctl_k_interface_name);
		printf("Interface: %s\n",ifname);
		
		port = lldpctl_get_port(iface);
		
		neighbors = lldpctl_atom_get(port, lldpctl_k_port_neighbors);

		//Iterate over each neighbor
		lldpctl_atom_foreach(neighbors,neigh){
			
			addresses = lldpctl_atom_get(neigh,lldpctl_k_chassis_mgmt);
			
			//Iterate over each management address
			lldpctl_atom_foreach(addresses,adr){			
				mgmt_ip = lldpctl_atom_get_str(adr,lldpctl_k_mgmt_ip);
				printf("mgmt_ip: %s\n",mgmt_ip);
				
			}//end foreach addresses
			lldpctl_atom_dec_ref(adr);
		
		}//end foreach neighbors
		lldpctl_atom_dec_ref(neigh);	
		
	}//end foreach interfcaes
	lldpctl_atom_dec_ref(iface);
```

**Comment by vincentbernat** [comment, marked as answer]
`lldpcli` uses the same API. So, you can look at its code to see how it is implemented. `lldpctl_watch_callback2()` is used in `lldpcli`.

**Reply by vne147** [reply]
Hello, Vincent.  Thank you for the quick reply.

Are you referring to `src/client/lldpcli.c`?

If so I looked through that file and didn't see any references to `lldpctl_watch_callback2()`.

I'm kind of a noob.  What am I doing wrong?

**Reply by vne147** [reply]
Never mind.  I found the relevant code blocks in `src/client/show.c`.  Thanks for the help!

**Comment by Ekmansoft** [comment]
We are using this API to gather info on POE status, and to reconfigure POE. However, we have a problem that during boot, lldpd is reconfigured twice with added device name and (it seems) port filtering. When this happens, it seems like the lldpd API freezes, and may give bad responses. 
It does not seem possible to get information (through any callback or similar) in the API that a reconfiguration has made lldpd API stop working. 

I am thinking of the time between these prints:

```
lldpcli[1344]: system name set to new value blablabla1
lldpcli[1344]: iface-pattern set to new value eth*,!eth1
lldpcli[1344]: description set to new value blablabla2
lldpcli[1344]: lldpd should resume operations

```


We have been discussing that maybe if we control everything from the process that uses the lldpd API, we may get rid of these problems since we can control when it is done. But this will cause a bit of work which we haven't started on yet. 

What is your proposal for these cases @vincentbernat?
